README
NAME

    Dist::Zilla::Plugin::MathInt64 - Include the Math::Int64 C client API
    in your distribution

VERSION

    version 0.07

SYNOPSIS

    in your dist.ini

     [PPPort]
     [MathInt64]
     [ModuleBuild]
     mb_class = MyDist::ModuleBuild

    in your xs (lib/MyDist.xs):

     #include "EXTERN.h"
     #include "perl.h"
     #include "XSUB.h"
     #include "ppport.h"
     
     /* provides int64_t and uint64_t if not   *
      * already available                      */
     #include "perl_math_int64_types.h"
     
     /* #define MATH_INT64_NATIVE_IF_AVAILABLE */
     #include "perl_math_int64.h"
     
     MODULE = MyDist  PACKAGE = MyDist
     
     int64_t
     function_that_returns_64bit_integer()
     
     void
     function_that_takes_64bit_integer(number)
         int64_t number
     
     SV *
     same_idea_but_with_xs(sv_number)
         SV *sv_number
       CODE:
         int64_t native_number = SvI64(sv_number);
         ...
         RETVAL = newSVi64(native_number);
       OUTPUT:
         RETVAL

    See Math::Int64#C-API for details.

    in your Module::Build subclass (inc/MyDist/ModuleBuild.pm):

     package MyDist::ModuleBuild;
     
     use base qw( Module::Build );
     
     sub new
     {
       my($class, %args) = @_;
       $args{c_source} = '.';
       $class->SUPER::new(%args);
     }

DESCRIPTION

    Math::Int64 provides an API for Perl and XS modules for dealing with 64
    bit integers.

    This plugin imports the C client API from Math::Int64 into your
    distribution. The C client API depends on ppport.h, so make sure that
    you also get that (the easiest way is via the PPPort plugin.

    This plugin will also create an appropriate typemap or update an
    existing typemap to automatically support the types int64_t and
    uint64_t in your XS code. (You can turn this off by setting typemap =
    0).

    This plugin will also declare Math::Int64 as a prerequisite for your
    distribution.

    One thing this plugin does NOT do is, it doesn't tell either
    Module::Build or ExtUtils::MakeMaker where to find the C and XS
    sources. One way of doing this would be to create your own
    Module::Build subclass and set the c_source attribute to where the C
    header and source code go (see the synopsis above as an example).

ATTRIBUTES

 dir

    Directory to dump the C source and header files into. If not specified,
    they go into the distribution root. If you use this option you probably
    need to tell the PPPort plugin to put the ppport.h file in the same
    place.

     [PPPort]
     filename = xs/ppport.h
     [MathInt64]
     dir = xs

 typemap

    If set to true (the default), then create a typemap file if it does not
    already exist with the appropriate typemaps for 64 bit integers, or if
    a typemap already exists, add the 64 bit integer mappings.

 typemap_path

    The path to the typemap file (if typemap is true). The default is
    simply 'typemap'.

CAVEATS

    This plugin uses ExtUtils::Typemaps to munge the typemaps file, which
    strips any comments from the typemap file, but should be semantically
    identical. Versions prior to 0.05 did its own parsing but would retain
    comments.

BUNDLED SOFTWARE

    This distribution comes bundled with C source code placed in the public
    domain by Salvador Fandino.

    Thanks to Salvador Fandino for writing Math::Int64 and providing a XS /
    C Client API for other distribution authors.

AUTHOR

    Graham Ollis <plicease@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2013 by Graham Ollis.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Dist-Zilla-Plugin-MathInt64

0.07      2014-11-03 05:59:25 -0500
  - update API to Math::Int64 0.34 (now includes sv_seti64 and sv_setu64 macros)

0.06      2013-12-31 12:30:02 -0500
  - Support Perl 5.8
  - add optional header perl_math_int64_types.h for int64_t and uint64_t where they aren't otherwise provided

0.05      2013-12-11 02:47:23 -0500
  - added typemap_path option
  - switched to Module::Build

0.04      2013-11-26 11:30:32 -0500
  - documentation improvements
  - first CPAN release

0.03      2013-11-20 10:10:11 -0500
  - add Math::Int64 as a prereq of target dist

0.02      2013-11-19 12:20:14 -0500
  - create/update typemap

0.01      2013-11-19 10:59:02 -0500
  - initial version


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.021.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Include the Math::Int64 C client API in your distribution",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-MathInt64",
  "dist_version" => "0.07",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::MathInt64",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dist::Zilla" => 0,
    "ExtUtils::Typemaps" => 0,
    "File::ShareDir" => 0,
    "Moose" => 0,
    "perl" => "5.006"
  },
  "script_files" => [],
  "share_dir" => {
    "dist" => "share"
  },
  "test_requires" => {
    "Path::Class" => "0.26",
    "Test::More" => "0.94"
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.3601",
  "Path::Class" => "0.26",
  "Test::More" => "0.94"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;

use File::ShareDir::Install;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Include the Math::Int64 C client API in your distribution",
  "AUTHOR" => "Graham Ollis <plicease\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "File::ShareDir::Install" => "0.03"
  },
  "DISTNAME" => "Dist-Zilla-Plugin-MathInt64",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Dist::Zilla::Plugin::MathInt64",
  "PREREQ_PM" => {
    "Dist::Zilla::File::InMemory" => 0,
    "Dist::Zilla::Role::FileGatherer" => 0,
    "Dist::Zilla::Role::FileMunger" => 0,
    "Dist::Zilla::Role::Plugin" => 0,
    "Dist::Zilla::Role::PrereqSource" => 0,
    "File::ShareDir" => 0,
    "Moose" => 0,
    "Path::Class::Dir" => 0,
    "Path::Class::File" => 0
  },
  "TEST_REQUIRES" => {
    "Path::Class" => 0,
    "Test::DZil" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dist::Zilla::File::InMemory" => 0,
  "Dist::Zilla::Role::FileGatherer" => 0,
  "Dist::Zilla::Role::FileMunger" => 0,
  "Dist::Zilla::Role::Plugin" => 0,
  "Dist::Zilla::Role::PrereqSource" => 0,
  "File::ShareDir" => 0,
  "Moose" => 0,
  "Path::Class" => 0,
  "Path::Class::Dir" => 0,
  "Path::Class::File" => 0,
  "Test::DZil" => 0,
  "Test::More" => "0.88"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}




share/perl_math_int64.h
/*
 * perl_math_int64.h - This file is in the public domain
 * Author: Salvador Fandino <sfandino@yahoo.com>
 * Version: 2.1
 *
 * Generated on: 2014-11-03 05:59:26
 * Math::Int64 version: 0.34
 * Module::CAPIMaker version: 
 */

#if !defined (PERL_MATH_INT64_H_INCLUDED)
#define PERL_MATH_INT64_H_INCLUDED

#define MATH_INT64_C_API_REQUIRED_VERSION 2
#define MATH_INT64_VERSION MATH_INT64_C_API_REQUIRED_VERSION

int perl_math_int64_load(int required_version);

#define PERL_MATH_INT64_LOAD perl_math_int64_load(MATH_INT64_C_API_REQUIRED_VERSION)
#define PERL_MATH_INT64_LOAD_OR_CROAK \
    if (PERL_MATH_INT64_LOAD);        \
    else croak(NULL);
#define MATH_INT64_BOOT PERL_MATH_INT64_LOAD_OR_CROAK

extern HV *math_int64_c_api_hash;
extern int math_int64_c_api_min_version;
extern int math_int64_c_api_max_version;
#define math_int64_capi_version math_int64_c_api_max_version

#if (defined(MATH_INT64_NATIVE_IF_AVAILABLE) && (IVSIZE == 8))
#define MATH_INT64_NATIVE 1
#endif

extern int64_t   (*math_int64_c_api_SvI64)(pTHX_ SV*);
#define SvI64(a) ((*math_int64_c_api_SvI64)(aTHX_ (a)))
extern int       (*math_int64_c_api_SvI64OK)(pTHX_ SV*);
#define SvI64OK(a) ((*math_int64_c_api_SvI64OK)(aTHX_ (a)))
extern uint64_t  (*math_int64_c_api_SvU64)(pTHX_ SV*);
#define SvU64(a) ((*math_int64_c_api_SvU64)(aTHX_ (a)))
extern int       (*math_int64_c_api_SvU64OK)(pTHX_ SV*);
#define SvU64OK(a) ((*math_int64_c_api_SvU64OK)(aTHX_ (a)))
extern SV *      (*math_int64_c_api_newSVi64)(pTHX_ int64_t);
#define newSVi64(a) ((*math_int64_c_api_newSVi64)(aTHX_ (a)))
extern SV *      (*math_int64_c_api_newSVu64)(pTHX_ uint64_t);
#define newSVu64(a) ((*math_int64_c_api_newSVu64)(aTHX_ (a)))
extern uint64_t  (*math_int64_c_api_randU64)(pTHX);
#define randU64() ((*math_int64_c_api_randU64)(aTHX))


#if MATH_INT64_NATIVE

#undef newSVi64
#define newSVi64 newSViv
#undef newSVu64
#define newSVu64 newSVuv

#define sv_seti64 sv_setiv_mg
#define sv_setu64 sv_setuv_mg

#else

#define sv_seti64(target, i64) (sv_setsv_mg(target, sv_2mortal(newSVi64(i64))))
#define sv_setu64(target, u64) (sv_setsv_mg(target, sv_2mortal(newSVu64(u64))))

#endif

#endif

share/perl_math_int64.c
/*
 * perl_math_int64.c - This file is in the public domain
 * Author: Salvador Fandino <sfandino@yahoo.com>
 *
 * Generated on: 2014-11-03 05:59:26
 * Math::Int64 version: 0.34
 * Module::CAPIMaker version: 
 */

#include "EXTERN.h"
#include "perl.h"
#include "ppport.h"

#ifdef __MINGW32__
#include <stdint.h>
#endif

#ifdef _MSC_VER
#include <stdlib.h>
typedef __int64 int64_t;
typedef unsigned __int64 uint64_t;
#endif

/* you may need to add a typemap for int64_t here if it is not defined
   by default in your C header files */

HV *math_int64_c_api_hash;
int math_int64_c_api_min_version;
int math_int64_c_api_max_version;

int64_t   (*math_int64_c_api_SvI64)(pTHX_ SV*);
int       (*math_int64_c_api_SvI64OK)(pTHX_ SV*);
uint64_t  (*math_int64_c_api_SvU64)(pTHX_ SV*);
int       (*math_int64_c_api_SvU64OK)(pTHX_ SV*);
SV *      (*math_int64_c_api_newSVi64)(pTHX_ int64_t);
SV *      (*math_int64_c_api_newSVu64)(pTHX_ uint64_t);
uint64_t  (*math_int64_c_api_randU64)(pTHX);

int
perl_math_int64_load(int required_version) {
    dTHX;
    SV **svp;
    eval_pv("require Math::Int64", TRUE);
    if (SvTRUE(ERRSV)) return 0;

   math_int64_c_api_hash = get_hv("Math::Int64::C_API", 0);
    if (!math_int64_c_api_hash) {
        sv_setpv(ERRSV, "Unable to load Math::Int64 C API");
        SvSETMAGIC(ERRSV);
        return 0;
    }

    svp = hv_fetch(math_int64_c_api_hash, "min_version", 11, 0);
    if (!svp) svp = hv_fetch(math_int64_c_api_hash, "version", 7, 1);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to retrieve C API version for Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_min_version = SvIV(*svp);

    svp = hv_fetch(math_int64_c_api_hash, "max_version", 11, 0);
    if (!svp) svp = hv_fetch(math_int64_c_api_hash, "version", 7, 1);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to retrieve C API version for Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_max_version = SvIV(*svp);

    if ((required_version < math_int64_c_api_min_version) ||
        (required_version > math_int64_c_api_max_version)) {
        sv_setpvf(ERRSV,
                  "Math::Int64 C API version mismatch. "
                  "The installed module supports versions %d to %d but %d is required",
                  math_int64_c_api_min_version,
                  math_int64_c_api_max_version,
                  required_version);
        SvSETMAGIC(ERRSV);
        return 0;
    }

    svp = hv_fetch(math_int64_c_api_hash, "SvI64", 5, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvI64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvI64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "SvI64OK", 7, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvI64OK' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvI64OK = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "SvU64", 5, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvU64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvU64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "SvU64OK", 7, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'SvU64OK' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_SvU64OK = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "newSVi64", 8, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'newSVi64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_newSVi64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "newSVu64", 8, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'newSVu64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_newSVu64 = INT2PTR(void *, SvIV(*svp));
    svp = hv_fetch(math_int64_c_api_hash, "randU64", 7, 0);
    if (!svp || !*svp) {
        sv_setpv(ERRSV, "Unable to fetch pointer 'randU64' C function from Math::Int64");
        SvSETMAGIC(ERRSV);
        return 0;
    }
    math_int64_c_api_randU64 = INT2PTR(void *, SvIV(*svp));

    return 1;
}
